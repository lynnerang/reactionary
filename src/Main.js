import React, { Component } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import './Main.scss';
import Home from './Home.js';
import Play from './Play.js';
import MyCards from './MyCards.js';
import Stats from './Stats.js';
import Links from './Links.js';

class Main extends Component {
  constructor(props) {
    super(props);

    this.state = {
      myCards: [],
      gameCount: 0,
      totalGuesses: 0,
      highScore: {totalCards: 0, guesses: 0}
    }
  }

  componentDidMount() {
    const myCards = JSON.parse(localStorage.getItem('my cards')) || [];
    const gameCount = JSON.parse(localStorage.getItem('game count')) || 0;
    const highScore = JSON.parse(localStorage.getItem('high score')) || {totalCards: 0, guesses: 0};
    const totalGuesses = JSON.parse(localStorage.getItem('guess count')) || 0;

    this.setState({
      myCards: myCards, 
      gameCount: gameCount, 
      highScore: highScore, 
      totalGuesses: totalGuesses
    });
  }

  updateGameCount = () => {
    const gameCount = this.state.gameCount + 1;
    this.setState({gameCount: gameCount});
    localStorage.setItem('game count', JSON.stringify(gameCount));
  }

  updateGuessCount = update => {
    let total = update === 'add' ? this.state.totalGuesses + 1 : 0;
    this.setState({totalGuesses: total});
    localStorage.setItem('guess count', JSON.stringify(total));
  }

  checkHighScore = (numCards) => {
    const numGuesses = this.state.totalGuesses;
    const highDiff =  this.state.highScore.guesses / this.state.highScore.totalCards;
    const currentDiff = numGuesses / numCards;

    if (!highDiff || currentDiff < highDiff) { 
      const newScore = {totalCards: numCards, guesses: numGuesses};
      this.setState({highScore: newScore}, () => {
        localStorage.setItem('high score', JSON.stringify(newScore));
      });
    } 
  }

  saveNewCard = (term, desc, file) => {
    const newCard = {id: Date.now(), term: term, definition: desc, example: file, new: true};
    const updatedCards = this.state.myCards;
    updatedCards.push(newCard);
    this.updateMyCards(updatedCards);
    this.updateStorageRemainingCards(newCard.id, 'add');
  }

  deleteMyCard = cardId => {
    const updatedCards = this.state.myCards; 
    const deletedCard = updatedCards.find(card => card.id === cardId);
    updatedCards.splice(updatedCards.indexOf(deletedCard), 1);
    this.updateMyCards(updatedCards);
    this.updateStorageRemainingCards(deletedCard.id);
  }

  updateMyCards = cards => {
    this.setState({myCards: cards});
    localStorage.setItem('my cards', JSON.stringify(cards))
  }

  updateStorageRemainingCards = (id, option) => {
    let remaining = JSON.parse(localStorage.getItem('remaining cards')) || [];
    if (option === 'add') {
      remaining.push(id);
    } else if (remaining.includes(id)) {
      remaining.splice(remaining.indexOf(id), 1);
    }
    localStorage.setItem('remaining cards', JSON.stringify(remaining));
  }

  markCardUsed = cardId => {
    const updatedCards = [...this.state.myCards];  
    const index = updatedCards.indexOf(updatedCards.find(card => card.id === cardId));
    updatedCards[index].new = false;
    this.updateMyCards(updatedCards);
  }

  render() {
    return (
        <main>
            <Switch>
              <Route path='/' exact render={(props) => (<Home />)}/>
              <Route path='/play' render={(props) => (
                <Play data={this.props.data} 
                      myCards={this.state.myCards}
                      updateGameCount={this.updateGameCount}
                      markCardUsed={this.markCardUsed}
                      checkHighScore={this.checkHighScore}
                      updateGuessCount={this.updateGuessCount}
                />
              )}/>
              <Route path='/mycards' render={(props) => (
                <MyCards data={this.props.data}
                        myCards={this.state.myCards}
                        saveNewCard={this.saveNewCard} 
                        deleteMyCard={this.deleteMyCard}
                />
              )}/>
              <Route path='/stats' render={(props) => (
                <Stats gameCount={this.state.gameCount}
                      highScore={this.state.highScore}
                />
              )}/>
              <Route path='/links' render={(props) => (<Links />)}/>
            </Switch>
        </main>
    )
  }
}

export default Main;