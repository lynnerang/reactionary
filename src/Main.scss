@import './variables.scss';

h2 {
  color: $blue;
  font-size: 38px;
  font-weight: 500;
  margin-bottom: 20px;
  letter-spacing: 3px;
}

h3 {
  color: $secondary-color;
  font-size: 24px;
  font-weight: 400;
  margin-bottom: 12px;
}

aside {
  min-height: calc(100vh - 70px);
}

.instructions {
  width: 96%;
  padding-left: 2px;
  font-size: 20px;
  line-height: 1.3;
  color: rgb(65, 65, 65);
  padding-bottom: 10px;
}

.hidden {
  display: none;
}