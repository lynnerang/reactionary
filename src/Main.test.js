import React from 'react';
import ReactDOM from 'react-dom';
import Main from './Main.js';
import { shallow } from 'enzyme';

const mockData = {
  id: 2,
  term: "Compilers",
  fakeAnswers: ["Bundlers", "Package Managers"],
  definition: "These things take JavaScript code, transform it, and return it as JavaScript code in a different format.",
  example: "Babel is a popular example, and is most commonly used for React and transforms ES6 syntax into syntax that older browsers can interpret."
}


describe('Main', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow (
      <Main 
        data={mockData}
      />
      )
  });

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has four default state properties', () => {
    expect(wrapper.state()).toEqual({
      myCards: [],
      gameCount: 0,
      totalGuesses: 0,
      highScore: {totalCards: 0, guesses: 0}
    });
  });

  it('can update the total game count', () => {
    expect(wrapper.state('gameCount')).toEqual(0);

    wrapper.instance().updateGameCount();

    expect(wrapper.state('gameCount')).toEqual(1);
  });

  it('can update a total guess count.', () => {
    expect(wrapper.state('totalGuesses')).toEqual(0);

    wrapper.instance().updateGuessCount('add');

    expect(wrapper.state('totalGuesses')).toEqual(1)
  });

  it('can check the current high score and update if it the arg is a higher score', () => {
    expect(wrapper.state('highScore')).toEqual({totalCards: 0, guesses: 0});

    wrapper.instance().updateGuessCount('add');

    wrapper.instance().checkHighScore(1);

    expect(wrapper.state('highScore')).toEqual({totalCards: 1, guesses: 2});
  });

  it('can save a new custom card to its myCards state.', () => {
    expect(wrapper.state('myCards').length).toEqual(0);

    wrapper.instance().saveNewCard('term1', 'def1', '');

    expect(wrapper.state('myCards').length).toEqual(1);
  });

  it('can delete a card from its myCards state.', () => {
    expect(wrapper.state('myCards').length).toEqual(1);

    wrapper.instance().saveNewCard('term2', 'def2', '');
    wrapper.instance().saveNewCard('term3', 'def3', '');

    expect(wrapper.state('myCards').length).toEqual(3);

    wrapper.instance().deleteMyCard(wrapper.state('myCards')[0].id);

    expect(wrapper.state('myCards').length).toEqual(2);
  });

  it('can update the myCards state to any array.', () => {
    expect(wrapper.state('myCards').length).toEqual(2);

    wrapper.instance().updateMyCards(['oh', 'hi', 'mark']);

    expect(wrapper.state('myCards')).toEqual(['oh', 'hi', 'mark']);
  });

  it('can set a custom card property of new to false.', () => {
    expect(wrapper.state('myCards')[0].new).toEqual(true);

    wrapper.instance().markCardUsed(wrapper.state('myCards')[0].id);

    expect(wrapper.state('myCards')[0].new).toEqual(false);
  });

  afterEach(() => {
    wrapper = null;
  });

});