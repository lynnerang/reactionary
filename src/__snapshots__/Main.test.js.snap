// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Should match the snapshot 1`] = `
<main>
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      path="/play"
      render={[Function]}
    />
    <Route
      path="/mycards"
      render={[Function]}
    />
    <Route
      path="/stats"
      render={[Function]}
    />
    <Route
      path="/links"
      render={[Function]}
    />
  </Switch>
</main>
`;
